deployment/ansible/scripts/topology.py

```diff
@@ -40,51 +45,132 @@ def hosts_to_peers(lines):
+                f"'./sharding-p2p-poc -loglevel=DEBUG "
+                f"-client addpeer {other.public_ip} {other.listen_port} {other.seed}'"
+            )
+        }
+        for item in topology
+        for other in item["connect_to"]
+    ]
+
+    return commands
+
+
+def to_subshard_commands(topology):
+    commands = [
+        {
+            "host": item["peer"].host,
+            "command": (
+                f"docker exec -t {item['peer'].container_name} sh -c "
+                f"'./sharding-p2p-poc -loglevel=DEBUG "
+                f"-client subshard {item['shard_id']}'")
+        }
+        for item in topology
+    ]
+    return commands
+
+
+def to_broadcastcollation_commands(topology):
+    commands = [
+        {
+            "host": item["peer"].host,
+            "command": (
+                f"docker exec -t {item['peer'].container_name} sh -c "
+                f"'./sharding-p2p-poc -loglevel=DEBUG "
+                f"-client broadcastcollation {item['shard_id']} {NUM_COLLATIONS} {COLLATION_SIZE} {FREQUENCY}'"
+            )
+        }
+        for item in topology if item["is_broadcasting"]
+    ]
+    return commands
+
+def visualize(topology):
+    import networkx as nx
+    G = nx.Graph()
+    for item in topology:
+        me = item["peer"].container_name
+        G.add_node(me, shard_id = item["shard_id"])
+        for other in item['connect_to']:
+            G.add_edge(me, other.container_name)
+    import matplotlib.pyplot as plt
+    pos_nodes = nx.spring_layout(G)
+    nx.draw(G, pos_nodes, with_labels=True)
+
+    pos_attrs = {node: (coords[0], coords[1] + 0.08) for node, coords in pos_nodes.items()}
+
+    node_attrs = nx.get_node_attributes(G, 'shard_id')
+    custom_node_attrs = {node: attr for node, attr in node_attrs.items()}
+    nx.draw_networkx_labels(G, pos_attrs, labels=custom_node_attrs)
+
+    plt.savefig("artifacts/network.png")
 
 
 if __name__ == '__main__':
     inventories = read_inventories(sys.argv[1])
-    topology = barbell_topology(inventories)
-    for p, l in topology.items():
-        print(p.container_name, p.public_ip, l)
-    print("\n")
-    yml = topology_to_yaml(topology)
+    nodes = inventories["[nodes]"]
+    hosts = list(map(lambda x: x.split(" ")[0], nodes))
+    print(hosts)
+    peers = hosts_to_peers(nodes)
+
+    print("\n--- [ Network Nodes ] ---\n")
+
+    for peer in peers:
+        print(peer.container_name, peer.public_ip)
+
+    topology = barbell_topology(peers)
+    functions = {
+        "addpeer": to_addpeer_commands,
+        "subshard": to_subshard_commands,
+        "broadcastcollation": to_broadcastcollation_commands
+    }
+    results = {
+        k: groupby_host(v(topology), hosts)
+        for k, v in functions.items()
+    }
+
+    yml = yaml.dump(results, default_flow_style=False, width=1000)
+
+    print("\n--- [ Commands ] ---\n")
     print(yml)
-    with open("topology.yml", "w") as f:
+    with open("artifacts/commands.yml", "w") as f:
```

Comment:
```bash
Traceback (most recent call last):
  File "scripts/topology.py", line 173, in <module>
    with open("artifacts/commands.yml", "w") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'artifacts/commands.yml'
```

Reply:
Oops, I'll add a `.gitkeep` for artifacts dir